### PR TITLE
Add functionality to add files from a list of paths

### PR DESCRIPTION
**Pull request by CharlesFrankum**
## Related Issue \ discussion
https://github.com/allegroai/clearml/issues/1127

## Patch Description
Adds logic to `_add_files` so that we can now additionally pass in a list of paths

## Testing Instructions
```
from os import listdir
from os.path import isfile, join

from clearml import Dataset


test_dir = ''
file_paths = [f for f in listdir(test_dir) if isfile(join(test_dir , f))]
dataset.add_files(paths=paths)
```

### CODE REVIEW

**Comment by eugen-ajechiloae-clearml** [comment]
Hi @CharlesFrankum ! I don't think we should have a new function here, but instead make `path` in `add_files` support sequences as well, similarly to how it's done in `add_external_files`: https://github.com/allegroai/clearml/blob/18210ed144ce630fb1bda44d07cea356a7e23b53/clearml/datasets/dataset.py#L436

**Comment by CharlesFrankum** [comment]
> Hi @CharlesFrankum ! I don't think we should have a new function here, but instead make `path` in `add_files` support sequences as well, similarly to how it's done in `add_external_files`:
> 
> https://github.com/allegroai/clearml/blob/18210ed144ce630fb1bda44d07cea356a7e23b53/clearml/datasets/dataset.py#L436

Ah ok! I wasn't sure if we wanted to accept multiple types here but makes sense - will make the changes 👍

**Comment by jkhenning** [comment]
Hi @CharlesFrankum , any update?

File: clearml/datasets/dataset.py [hunk 1]
```diff
@@ -1975,39 +1975,50 @@ def _add_files(
...
+                    file_entries.extend([f for f in files if f.is_file()])
+                file_entries = list(set(file_entries))
+                file_entries = [
+                    FileEntry(
+                        parent_dataset_id=self._id,
+                        local_path=f.absolute().as_posix(),
+                        relative_path=(Path(dataset_path or ".") / f.relative_to(local_base_folder)).as_posix(),
+                    )
+                    for f in file_entries
+                ]
```
Comment (eugen-ajechiloae-clearml) [inline clearml/datasets/dataset.py:2013]: let's have this section as a different function, called `def _resolve_file_entries`. Don't calculate the hashes in this function.

File: clearml/datasets/dataset.py [hunk 2]
```diff
@@ -1975,39 +1975,50 @@ def _add_files(
...
-            file_entries = [
-                FileEntry(
-                    parent_dataset_id=self._id,
-                    local_path=f.absolute().as_posix(),
-                    relative_path=(Path(dataset_path or ".") / f.relative_to(local_base_folder)).as_posix(),
-                )
-                for f in file_entries
-            ]
+            files = [Path(p) for p in path if Path(p).exists()]
+            file_entries = [f for f in files if f.is_file()]
```
Comment (eugen-ajechiloae-clearml) [inline clearml/datasets/dataset.py:2018]: Here, you should call the `_resolve_file_entries` on each path specified in `path`:
```py
if isinstance(path, (str, Path, _Path)):
    path = [path]  # ensure list
file_entries = []
for p in path:
    file_entries.extend(self._resolve_file_entries(p))

# now calculate the hashes
```

**Review by eugen-ajechiloae-clearml — CHANGES_REQUESTED** [review]
I think we can improve this a bit :+1:

File: clearml/datasets/dataset.py [hunk 3]
```diff
@@ -1948,6 +1948,56 @@ def list_datasets(
...
+        :param local_base_folder: files will be located based on their relative path from local_base_folder
+        :param dataset_path: where in the dataset the folder/files should be located
+        :param recursive: If True, match all wildcard files recursively
+        """
+        path = Path(path)
+        local_base_folder = Path(local_base_folder or path)
+        if path.is_file():
+            if not local_base_folder.is_dir():
+                local_base_folder = local_base_folder.parent
+            file_entry = self._calc_file_hash(
```
Comment (eugen-ajechiloae-clearml) [inline clearml/datasets/dataset.py:1975]: No need to calculate the hash here

**Review by eugen-ajechiloae-clearml — CHANGES_REQUESTED** [review]

**Comment by eugen-ajechiloae-clearml** [comment]
Hi @CharlesFrankum ! Why was this PR closed? Any way we can help reopen it?

**Comment by CharlesFrankum** [comment]
> Hi @CharlesFrankum ! Why was this PR closed? Any way we can help reopen it?

Ah sorry, I messed up my master branch on the fork so reset it so I could work on a different PR and it automatically closed this PR. I can reopen it and work on it in a little. Sorry for never finishing this one btw - we ended up moving away from clearml datasets for now so was low priority

**Comment by eugen-ajechiloae-clearml** [comment]
@CharlesFrankum It would be great if we could have this PR merged. If you could reopen it that would be great. If you don't wish to work on it anymore I can take over.

**Pull request closed without merging**